content/posts/building_jnigen.md

```diff
@@ -0,0 +1,262 @@
+
+Surprisingly, most platform classes are not available if you call JNI's `FindClass` from a flutter thread. This is because those threads have a much barebones classloader. The solution suggested by [Android Developers documentation](https://developer.android.com/training/articles/perf-jni) is to store a reference to original thread's class loader and call its loadClass method instead of JNI's `FindClass`, which worked for us.
+
+### The perils of dynamic loading
+The generated C code needs to call the function in C support library. At minimum, it should be able to access the shared context such as the class loader. However, we can't link them in compile time, with existing Flutter build system.
+
+The other solution is to load the DLLs into a shared namespace. However it was not feasible because on some platforms, `DynamicLibrary.open` implementation loads symbols into local namespaces of the library, and not expose it. [^rtld]. We worked around it by having the initialization code in dart, which set some pointers in generated C code. [^why_pure_dart]
+
+After all the workarounds, it was possible to use JNI from the support library `package:jni`. Although it wasn't most ergonomic thing to use directly, it was supposed to provide a layer shared by all generated wrappers.
+
+### A glimpse of one-off APIs
+My litmus test at these stages was being able to call some Android built-in APIs, such as displaying a `Toast` message, or getting time since boot [^real_world_examples]. The latter will be a simple one-liner on Android. 
+
+```java
+long millisecondsSinceBoot = SystemClock.uptimeMillis();
+```
+
+If you call this using JNI primitives, there will be a number of steps
+
+* Find class `SystemClock`.
+* Find static method `uptimeMillis` on `SystemClock`. A fairly cryptic signature also needs to be specified. Here the signature is `()J` which means this method takes no arguments and returns a `long`.
+* Call this static method and obtain the result.
+* Check if there's an exception.
+* Return the result to the caller.
+
+While the ideal way to avoid this is generating wrapper, support library `package:jni` still provides a way to call methods without generating code. This layer of abstraction is intended for _one-off usages and testing_. It's still quite verbose but better than calling the methods on `JNIEnv` directly.
+
+```dart
+Jni.findJClass("android/os/SystemClock").use(
+    (systemClock) => systemClock.callStaticMethodByName<int>(
+        "uptimeMillis", "()J", [], JniCallType.longType),
+  );
+```
+
+These are "stringy" APIs which are only supposed to be for one-off uses where you can't generate code, and debugging. The actual `jnigen`-generated bindings will use slightly lower level APIs, storing method references etc.. into class fields when appropriate.
+
+[Screenshot of an Example app Demonstrating various JNI calls](images/building_jnigen/Dart_JNI_Example_Screenshot.png)
+
+## Parsing Java libraries
+
+### The problem: we need an AST
```

Comment:
nit: Some headers have an empty line after them, some don't.

Reply:
Level-2 headers have empty line, Level-3 doesn't. Shouldn't matter in final blog though.